python/tests/test_tables.py

```diff
@@ -255,12 +237,87 @@ def test_set_read_only_attributes(self):
+            table.set_columns(**input_data)
+
+            for list_col, offset_col in self.ragged_list_columns:
+                list_data = input_data[list_col.name]
+                self.assertTrue(
+                    np.array_equal(getattr(table, list_col.name), list_data))
+                list_data += 1
+                self.assertFalse(
+                    np.array_equal(getattr(table, list_col.name), list_data))
+                setattr(table, list_col.name, list_data)
+                self.assertTrue(
+                    np.array_equal(getattr(table, list_col.name), list_data))
+                list_value = getattr(table[0], list_col.name)
+                self.assertEqual(len(list_value), 1)
+
+                # Reset the offsets so that all the full array is associated with the
+                # first element.
+                offset_data = np.zeros(num_rows + 1, dtype=np.uint32) + num_rows
+                offset_data[0] = 0
+                setattr(table, offset_col.name, offset_data)
+                list_value = getattr(table[0], list_col.name)
+                self.assertEqual(len(list_value), num_rows)
+
+                del input_data[list_col.name]
+                del input_data[offset_col.name]
+
+            for col, data in input_data.items():
+                self.assertTrue(np.array_equal(getattr(table, col), data))
+                data += 1
+                self.assertFalse(np.array_equal(getattr(table, col), data))
+                setattr(table, col, data)
+                self.assertTrue(np.array_equal(getattr(table, col), data))
+
+    def test_set_column_attributes_errors(self):
+        table = self.table_class()
+        num_rows = 10
+        input_data = {col.name: col.get_input(num_rows) for col in self.columns}
+        for list_col, offset_col in self.ragged_list_columns:
+            value = list_col.get_input(num_rows)
+            input_data[list_col.name] = value
+            input_data[offset_col.name] = np.arange(num_rows + 1, dtype=np.uint32)
+        table.set_columns(**input_data)
+
+        for list_col, offset_col in self.ragged_list_columns:
+            for bad_list_col in [[], input_data[list_col.name][:-1]]:
+                with self.assertRaises(ValueError):
+                    setattr(table, list_col.name, bad_list_col)
+            for bad_offset_col in [[], np.arange(num_rows + 2, dtype=np.uint32)]:
+                with self.assertRaises(ValueError):
+                    setattr(table, offset_col.name, bad_offset_col)
+
+            del input_data[list_col.name]
+            del input_data[offset_col.name]
+
+        for col, data in input_data.items():
+            for bad_data in [[], data[:-1]]:
+                with self.assertRaises(ValueError):
+                    setattr(table, col, bad_data)
+
+        # Try to access a column that isn't there.
+        with self.assertRaises(AttributeError):
```

Comment:
Should we try getting and setting separately - if you test both together like this then one could not raise an error whereas the other could, and you wouldn't spot it.

Reply:
We can always set a new attribute on Python objects, so ``table.no_such_column = 1`` will not raise an error. I've update the tests to make it clear that we're just failing on the read side though.

Reply:
Oh, good point